ksql-engine/src/main/java/io/confluent/ksql/function/UdfLoader.java

```diff
@@ -0,0 +1,154 @@
+  private final UdfCompiler compiler;
+  private final boolean loadCustomerUdfs;
+
+
+  public UdfLoader(final MetaStore metaStore,
+                   final File pluginDir,
+                   final ClassLoader parentClassLoader,
+                   final Predicate<String> blacklist,
+                   final UdfCompiler compiler,
+                   final boolean loadCustomerUdfs) {
+    this.metaStore = Objects.requireNonNull(metaStore, "metaStore can't be null");
+    this.pluginDir = Objects.requireNonNull(pluginDir, "pluginDir can't be null");
+    this.parentClassLoader = Objects.requireNonNull(parentClassLoader,
+        "parentClassLoader can't be null");
+    this.blacklist = Objects.requireNonNull(blacklist, "blacklist can't be null");
+    this.compiler = Objects.requireNonNull(compiler, "compiler can't be null");
+    this.loadCustomerUdfs = loadCustomerUdfs;
+  }
+
+  public void load() {
+    // load udfs packaged as part of ksql first
+    loadUdfs(parentClassLoader);
+    if (loadCustomerUdfs) {
+      try {
+        Files.find(pluginDir.toPath(), 1, (path, attributes) -> path.toString().endsWith(".jar"))
+            .map(path -> UdfClassLoader.newClassLoader(path, parentClassLoader, blacklist))
+            .forEach(this::loadUdfs);
+      } catch (IOException e) {
+        logger.error("Failed to load UDFs from location {}", pluginDir, e);
+      }
+    }
+  }
+
+
+  private void loadUdfs(final ClassLoader loader) {
+    new FastClasspathScanner()
+        .overrideClassLoaders(loader)
+        .ignoreParentClassLoaders()
+        // if we are loading from the parent classloader then restrict the name space to only
+        // jars/dirs containing "ksql-engine". This is so we don't end up scanning every jar
+        .filterClasspathElements(
+            name -> parentClassLoader != loader || name.contains("ksql-engine"))
+        .matchClassesWithMethodAnnotation(Udf.class,
+            (theClass, executable) -> {
+              final UdfDescription annotation = theClass.getAnnotation(UdfDescription.class);
+              if (annotation != null) {
+                final Method method = (Method) executable;
+                try {
+                  final UdfInvoker udf = compiler.compile(method, loader);
+                  addFunction(annotation, method, udf);
+                } catch (final KsqlException e) {
+                  if (parentClassLoader == loader) {
+                    throw e;
+                  } else {
+                    logger.warn("Failed to add UDF to the MetaStore. name={} method={}",
+                        annotation.name(),
+                        method,
+                        e);
+                  }
+                }
+              }
+            })
+        .scan();
+  }
+
+  private void addFunction(final UdfDescription annotation,
+                           final Method method,
+                           final UdfInvoker udf) {
+    metaStore.addFunction(new KsqlFunction(
```

Comment:
Is it worth calling `method.getDeclaringClass().newInstance()` once before adding the function to ensure that, in principle, instances can be created.